server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -165,7 +185,8 @@ public TransportVersion read(StreamInput in, String key) throws IOException {
 
     private final Map<String, TransportVersion> transportVersions;
     private final TransportVersion minTransportVersion;
-
+    private final Map<String, VersionsWrapper> versionsWrappers;
+    private final VersionsWrapper minVersions;
```

Comment:
Could we first generalize this per-node `TransportVersion` map into a `VersionsWrapper` map (containing just the `TransportVersion`) and then add the system index version to `VersionsWrapper` in a separate step? I believe that first step should be possible without changing the transport protocol.

Reply:
I've made a refactoring PR that just wraps TransportVersion with no changes to the transport protocol: https://github.com/elastic/elasticsearch/pull/99114